qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java

```diff
@@ -0,0 +1,265 @@
+            indexRandomDocuments(index, count, true);
+            createSnapshot();
+            // Explicitly flush so we're sure to have a bunch of documents in the Lucene index
+            client().performRequest("POST", "/_flush");
+            if (shouldHaveTranslog) {
+                // Update a few documents so we are sure to have a translog
+                indexRandomDocuments(index, count / 10, false /* Flushing here would invalidate the whole thing....*/);
+            }
+
+            // Record how many documents we built so we can compare later
+            XContentBuilder infoDoc = JsonXContent.contentBuilder().startObject();
+            infoDoc.field("count", count);
+            infoDoc.field("should_have_translog", shouldHaveTranslog);
+            infoDoc.endObject();
+            client().performRequest("PUT", infoDocument, singletonMap("refresh", "true"),
+                    new StringEntity(infoDoc.string(), ContentType.APPLICATION_JSON));
+        } else {
+            // Load the number of documents that were written to the old cluster
+            String doc = EntityUtils.toString(
+                    client().performRequest("GET", infoDocument, singletonMap("filter_path", "_source")).getEntity());
+            Matcher m = Pattern.compile("\"count\":(\\d+)").matcher(doc);
+            assertTrue(doc, m.find());
+            count = Integer.parseInt(m.group(1));
+            m = Pattern.compile("\"should_have_translog\":(true|false)").matcher(doc);
+            assertTrue(doc, m.find());
+            shouldHaveTranslog = Booleans.parseBoolean(m.group(1));
+        }
+
+        // Count the documents in the index to make sure we have as many as we put there
+        String countResponse = EntityUtils.toString(
+                client().performRequest("GET", "/" + index + "/_search", singletonMap("size", "0")).getEntity());
+        assertThat(countResponse, containsString("\"total\":" + count));
+
+        if (false == runningAgainstOldCluster) {
+            assertTranslogRecoveryStatistics(index, shouldHaveTranslog);
+        }
+
+        restoreSnapshot(index, count);
+
+        // TODO finish adding tests for the things in OldIndexBackwardsCompatibilityIT
+    }
+
+    // TODO tests for upgrades after shrink. We've had trouble with shrink in the past.
+
+    private void indexRandomDocuments(String index, int count, boolean flushAllowed) throws IOException {
+        for (int i = 0; i < count; i++) {
+            XContentBuilder doc = JsonXContent.contentBuilder().startObject(); {
+                doc.field("string", randomAlphaOfLength(10));
+                doc.field("int", randomInt(100));
+                doc.field("float", randomFloat());
+                // be sure to create a "proper" boolean (True, False) for the first document so that automapping is correct
+                doc.field("bool", i > 0 && supportsLenientBooleans ? randomLenientBoolean() : randomBoolean());
```

Comment:
I don't think we need to test lenient booleans. Sure the old cluster supported them, but they supported proper booleans too. They are gone from master now, so let's just stick with keeping this simple, there is enough complicated stuff here.